### DISCUSSION TITLE
Windows Compatibility

### DISCUSSION THREAD

**Discussion by S0UL4** [obfusk/apksigcopier#101] [Q&A]
Hello everyone,

I have a question that might seem simple, but I'm encountering an issue. When I run a Python script on Ubuntu, it successfully signs an unsigned app. However, when I switch to Windows, although I can generate the output APK, it remains unsigned.

Could anyone provide some assistance?

Thank you!

**Comment by obfusk** [comment]
Please provide sufficient information to allow providing assistance and confirm you read "Please note the following" in #96, specifically that you understand `apksigcopier` is a tool for reproducible builds only and that you are trying to use it as such.

> When I run a Python script on Ubuntu

What Python script? You mean `apksigcopier`?

> it successfully signs an unsigned app.

Unlike e.g. `apksigner`, `apksigcopier` doesn't sign, it copies signatures for reproducible builds. Are you sure you're using the right command?

> However, when I switch to Windows, although I can generate the output APK, it remains unsigned.

What output APK? What steps did you take to generate it? What commands did you use?

**Reply by S0UL4** [reply]
Sorry for the incomplete information.

Actually, I created a GUI using Tkinter ( with python ) to utilize the functionalities of apksigcopier, such as patch, copy, etc. so i'm using the python API of apksigcopier.

To copy the signature from a signed app to an unsigned app, I should use the do_copy method. However, when I execute this function on Windows, for example do_copy(signed_apk, unsigned_apk, output_apk, v1_only=False), I get the output APK (which is supposed to have the same signature as the signed APK). However, when I check it with apksigtool or apksigner, I find that the signatures didn't get copied, and the output of apksigner indicates that the signatures (v2, v3) are not verified. Conversely, when I perform exactly the same steps ( usually via the python API ) to copy the signatures from the signed app to the unsigned app on Linux, I get the output APK, and upon checking it with apksigner, it tells me that it has the copied signatures (v2 and v3)..

**Reply by obfusk** [reply]
> that the signatures (v2, v3) are not verified

Are the signatures missing? Or are they present but fail to verify? Can you provide e.g. `apksigner` output?

Can you try copying the signature from `golden-aligned-v1v2v3-out.apk` to `golden-aligned-in.apk` -- found [here](https://github.com/obfusk/test-apks) -- on Windows and provide the output APK?

Also: what Python version are you using?

**Reply by S0UL4** [reply]
Thank you for your quick reply.

Python used: 3.10.11

here is the output file (for copying the signature from golden-aligned-v1v2v3-out.apk to golden-aligned-in.apk  ) : https://easyupload.io/8sireu

Will provide also the output of apksigner.

**Reply by S0UL4** [reply]
The apksigtool actually output: 'v2 verified (1 signer(s)), v3 verified (1 signer(s))'; when testing with the above APKs ( golden-aligned-v1v2v3-out.apk to golden-aligned-in.apk )

I think it worked. However, copying the signature from 'golden-pinsapp-signed.apk' to 'pinsapp-unsigned.apk' did not work

**Reply by obfusk** [reply]
Please simply attach files here instead of linking to some third party website.

> copying the signature from 'golden-pinsapp-signed.apk' to 'pinsapp-unsigned.apk' did not work

That's expected. Those are not identical (apart from the signature). The signed one contains an extra file.

> I think it worked.

And using the exact same files that didn't work on Windows did work on Ubuntu?

I don't think I can debug this without more information about (or a copy of) the APKs you're using. I also don't have access to Windows to test anything myself. But AFAIK everything should work fine on Windows. I know people have used it successfully on Windows before.

**Reply by S0UL4** [reply]
Thank you so much. I will let you know if anything is not working correctly on Windows.

Regarding the testing, when I copy the sig from golden-aligned-v1v2v3-out.apk to golden-aligned-in.apk and drag the output apk in an Android device to test it, it is not openable. However, I can open it on an emulator.

**Reply by obfusk** [reply]
> Regarding the testing, when I copy the sig from golden-aligned-v1v2v3-out.apk to golden-aligned-in.apk and drag the output apk in an Android device to test it, it is not openable. However, I can open it on an emulator.

Those APKs are part of the `apksigner` (and `apksigcopier`) test suite, not actually functional apps AFAIK.

> And using the exact same files that didn't work on Windows did work on Ubuntu?

Or did you e.g. build one unsigned APK on Ubuntu and the other on Windows, meaning the issues could have been caused by different build outputs on those platforms?

If you can provide me with a pair of signed and unsigned APKs (and the output APK) for which copying works correctly on Ubuntu but fails on Windows -- using the _exact same files_ -- I can investigate what might be the cause. But I would need actual APKs to investigate.
